src/puppet.ts

```diff
@@ -378,8 +362,71 @@ export abstract class Puppet extends PuppetEventEmitter {
    *
    *
    */
-  public abstract async start () : Promise<void>
-  public abstract async stop ()  : Promise<void>
+  public async start () : Promise<void> {
+    this.on('heartbeat', this.feedDog)
+    this.watchdog.on('reset', this.dogReset)
+
+    /**
+     * 2. Setup `reset` Event via a 1 second Throttle Queue:
+     */
+    this.resetThrottleQueue = new ThrottleQueue<string>(1000)
+    // 2.2. handle all `reset` events via the resetThrottleQueue
+    this.on('reset', this.throttleReset)
+    // 2.3. call reset() and then ignore the following `reset` event for 1 second
+    this.resetThrottleQueueSubscription = this.resetThrottleQueue.subscribe(reason => {
+      log.silly('Puppet', 'constructor() resetThrottleQueue.subscribe() reason: "%s"', reason)
+      this.reset(reason)
+    })
+  }
+
+  public async stop (): Promise<void> {
+    this.removeListener('heartbeat', this.feedDog)
```

Comment:
How about let's leave the listener on our watchdog because I believe there is enough for us in stop() for just making the dog sleep?

Please correct me if there's something else we should take care of.

Reply:
If we do like this way, we are confident that, no matter how the `wechaty-puppet-xxx` operates the watchdog, there won't be problem after we reset the `wechaty-puppet`, otherwise, we can not recover the status of the `watchdog` is someone change it for some weird reason.

Reply:
Yes, that's the reason why the `watchdog` is private before.

After this PR has been landed, I believe we can make our `watchdog` private again, so that we can make sure there's no one else could feed the watchdog.

Do you feel comfortable with that?

Reply:
Yep, I agree with that.

We can move this to constructor, won't insist, just a different thought.

Reply:
Thanks for the explanation. I'm ok with it.